getlatest.go

```diff
@@ -0,0 +1,107 @@
+package namesys
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"time"
+
+	ggio "github.com/gogo/protobuf/io"
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	pb "github.com/libp2p/go-libp2p-pubsub-router/pb"
+)
+
+type getLatestProtocol struct {
+	host host.Host
+}
+
+func newGetLatestProtocol(host host.Host, getLocal func(key string) ([]byte, error)) *getLatestProtocol {
+	p := &getLatestProtocol{host}
+
+	host.SetStreamHandler(PSGetLatestProto, func(s network.Stream) {
+		p.Receive(s, getLocal)
+	})
+
+	return p
+}
+
+func (p *getLatestProtocol) Receive(s network.Stream, getLocal func(key string) ([]byte, error)) {
+	r := ggio.NewDelimitedReader(s, 1<<20)
+	msg := &pb.RequestLatest{}
+	if err := r.ReadMsg(msg); err != nil {
+		if err != io.EOF {
+			s.Reset()
+			log.Infof("error reading request from %s: %s", s.Conn().RemotePeer(), err)
+		} else {
+			// Just be nice. They probably won't read this
+			// but it doesn't hurt to send it.
+			s.Close()
+		}
+		return
+	}
+
+	response, err := getLocal(*msg.Identifier)
+	var respProto pb.RespondLatest
+
+	if err != nil || response == nil {
```

Comment:
Any thoughts on if I should treat errors that aren't `not found` style errors differently? Exiting the protocol nicely with just "nope I don't have any data for you" seems good to me.

Reply:
Discussed on slack: The plan is to have a status field that can be one of SUCCESS, NOT_FOUND, or ERR.